### DISCUSSION TITLE
Replace labstack/echo router.

### DISCUSSION THREAD

**Discussion by gobijan** [pocketbase/pocketbase#1480] [Polls]
Hi,

Pocketbase is a great platform and it is important to get it to a stable version 1 to get even more traction besides hobby projects.
Under the hood pocketbase is making use of the echo router in version 5, which is currently in alpha.

Personally I am not a fan of echo router because it is not standard Go http handler compliant but its version 5 being in alpha (see: https://github.com/labstack/echo/discussions/2000) is in my view kind of problematic.

There are other, more stable and standard compliant routers out there that offer similar features to echo like chi-router.
Any opinion on this? Would it be feasible to get rid of the echo router dependency?

PS: Having a standard compliant router would also open up to more tooling for OpenApi/Swagger.

Cheers Bijan

**Comment by ganigeorgiev** [comment]
There are no plans for now to replace echo.

As stated in the echo discussion, v5 is mostly a maintenance release to cleanup some v4 inconsistencies and I consider it stable enough. Additionally, we also have integration tests for most endpoints that should ensure compatibility for the important parts in case of a major change in echo.

I'm not entirely sure what do you mean by "_not standard compliant router_". You can pass a `http.Handler`, if that's what you want, by simply wrapping it with `echo.WrapHandler(handler)`.

**Reply by gobijan** [reply]
Thx for your quick reply.
Okay that's fine then I guess.
By standard compliant I meant exactly that. I am aware that it's possible to use the wrapper function or to just extract r and w from the context. It's just a thing in the Go community that "well behaved" routers/middlewares adhere to that standard.

**Reply by ganigeorgiev** [reply]
There will be always some trade offs when having a hard dependency, but in this specific case I think it is insignificant. Internally it is still the standard go `net/http` and it doesn't prevent users to integrate with 3rd party packages if they need to - `Get("/", handler)` vs `Get("/", echo.WrapHandler(handler))`.

**Reply by gobijan** [reply]
Agreed :) Is it still possible to use a different router in a subroute?
I tried to inject a chi-router as subrouter into a route but was unsuccessful doing so.
My goal was to make use of https://github.com/swaggest/rest which is a chi-router based clean architecture swagger framework.

**Reply by ganigeorgiev** [reply]
I'm not familiar with the package but I don't think that you can just "swap" directly one router with another, even if we used chi as router. Checking their [officially linked example](https://dev.to/vearutop/tutorial-developing-a-restful-api-with-go-json-schema-validation-and-openapi-docs-2490) where gin is used (they're replacing it with swaggest), they also have to do some manual work in order to integrate it and it seems that this will work only if you have access to all http handlers. It is kindof the reverse integration: PocketBase in swaggest/rest, not swaggest/rest in PocketBase.

Additionally, because PocketBase collection fields are defined during runtime and they can change from the UI, generating a proper OpenAPI docs will require some extra mapping and will not work out of the box with any generic swagger generator, if that's what you're trying to do. There is an open issue for swagger integration in #945, but for now it is a low priority.

**Comment by W1M0R** [comment]
It looks like Echo doesn't support SSE when setting up reverse proxy endpoints: https://github.com/labstack/echo/issues/1172

Am I correct in saying that if you want to proxy a custom SSE endpoint to another SSE server, then this wouldn't be possible in Pocketbase, due to the missing functionality in Echo?

**Reply by ganigeorgiev** [reply]
@W1M0R I don't understand what do you mean. The above linked issue seems related to the Echo proxy middleware, not to echo or net/http specifically.

In any case if your have other questions, please open a new Q&A discussion with more details about your use case.

Also keep in mind that the above comment from https://github.com/pocketbase/pocketbase/discussions/1480#discussioncomment-4580552 is no longer valid - see https://github.com/pocketbase/pocketbase/discussions/4355.

**Reply by W1M0R** [reply]
Thanks for your reply @ganigeorgiev. Based on your goals mentioned in the linked feature freeze issue, I don't think continuing this specific discussion will be a good use of time, and am happy to follow your new vision instead.
